### PR TITLE
feat: exclude emails and domains

### LINKED ISSUE #17580: ability to exclude emails and domains
CleanShot 2024-11-11 at 11 32 47@2x

comma separated:
CleanShot 2024-11-11 at 11 33 24@2x

content could be a full domain or a part of a domain

### PR DESCRIPTION
**Pull request by VK-RED**
## What does this PR do?

Added support for excluding emails and domains option in booking forms

- Fixes #17580 
- Fixes CAL-4697

/claim #17580 

[Screencast from 2024-11-12 09-33-33.webm](https://github.com/user-attachments/assets/3d1d55e8-812d-4f9c-9fd8-d25a723ab2e2)

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by emrysal** [comment]
Thinking about the wording here: It's not a feature that blocks the work email, it's any email that matches a given pattern. So "please try a work email" is not great here, better would be something like "Unfortunately, some email domains are not permitted. If your email was declined, please try a different email address."

**Review by emrysal — DISMISSED** [review]
We need some validation on the inputField to ensure the domains entered are valid and that the comma separated value is correct. @hariombalhara I also think the schema itself is very specific in this situation, maybe a more generic feature like excludedPattern is better - would like to know your opinion.

**Review by Praashh — DISMISSED** [review]
Can you please add lowercase checks to handle emails and domains?

**Comment by VK-RED** [comment]
Updating the changes

**Comment by PeerRich** [comment]
> Thinking about the wording here: It's not a feature that blocks the work email, it's any email that matches a given pattern. So "please try a work email" is not great here, better would be something like "Unfortunately, the email domain [domain].com is not permitted. Please try a different email address."

this requirement came from me and we should keep it this way for now.

Please move forward with "Please try a work email"

File: apps/web/public/static/locales/en/common.json [hunk 1]
```diff
@@ -2769,5 +2769,7 @@
   "salesforce_route_to_custom_lookup_field": "Route to a user that matches a lookup field on an account",
   "salesforce_option": "Salesforce Option",
   "lookup_field_name": "Lookup Field Name",
+  "exclude_emails_that_contain" : "Exclude emails that contain ...",
+  "exclude_emails_match_found_error_message" : "Unfortunately, the email domain is not permitted. Please try a different email address.",
```
Comment (PeerRich) [inline apps/web/public/static/locales/en/common.json:2773]:
```suggestion
  "exclude_emails_match_found_error_message" : "Please enter a valid work email address",
```

File: apps/web/public/static/locales/en/common.json [hunk 2]
```diff
@@ -2769,5 +2769,7 @@
   "salesforce_route_to_custom_lookup_field": "Route to a user that matches a lookup field on an account",
   "salesforce_option": "Salesforce Option",
   "lookup_field_name": "Lookup Field Name",
+  "exclude_emails_that_contain" : "Exclude emails that contain ...",
+  "exclude_emails_match_found_error_message" : "Please enter a valid work email address",
```
Comment (PeerRich) [inline apps/web/public/static/locales/en/common.json:2773]: this feature is primarily to throw this error: 
image

**Comment by VK-RED** [comment]
Added domain validation 

[Screencast from 2024-11-13 17-26-41.webm](https://github.com/user-attachments/assets/1131a583-e0a2-4ba7-8896-e9421513bd2b)

**Comment by Praashh** [comment]
Hey @VK-RED, thank you so much for your contributions, for now, we are going with the other one.

**Comment by Praashh** [comment]
Hey @VK-RED, thank you so much for your contributions, for now, we are going with the other one.

**Review by Praashh — CHANGES_REQUESTED** [review]
we need the correct version of the error for validation
### Current 
image
### Expected 
image

**Comment by VK-RED** [comment]
Hi @Praashh Everything seems to working fine for me. 

[Screencast from 2024-11-14 15-34-14.webm](https://github.com/user-attachments/assets/276bbf7a-d990-447c-bb00-3d2c9f40af28)

File: packages/features/form-builder/FormBuilder.tsx [hunk 3]
```diff
@@ -573,6 +574,16 @@ function FieldEditDialog({
                       <FieldWithLengthCheckSupport containerClassName="mt-6" fieldForm={fieldForm} />
                     ) : null}
 
+                    {/* render only for primary email field */}
```
Comment (Praashh) [inline packages/features/form-builder/FormBuilder.tsx:577]: please remove this.

**Review by Praashh — COMMENTED** [review]
Overall, it looks good to me. Please make the changes mentioned above, and I'll approve.

**Comment by VK-RED** [comment]
Hi @Praashh updated the changes

**Comment by Praashh** [comment]
> Hi @Praashh updated the changes

we also need unit test to verify this functionality, can you please add ?

**Comment by VK-RED** [comment]
Sure will add it !

**Comment by VK-RED** [comment]
Hi @Praashh added the tests

**Review by Praashh — APPROVED** [review]
LGTM!! thank you @VK-RED .

**Pull request merged**